### PR TITLE
feat(linter): implement `noUnknownTypeSelector` rule

### PR DESCRIPTION
**Pull request by Kazuhiro-Mimaki**
Closes https://github.com/biomejs/biome/issues/2543

## Summary
implements [selector-type-no-unknown](https://stylelint.io/user-guide/rules/selector-type-no-unknown/)

## Test Plan
Add snapshots.

### CODE REVIEW

File: crates/biome_css_analyze/src/lint/nursery/no_unknown_type_selector.rs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    ///
+    /// ```css
+    /// x-foo {}
+    /// ```
+    ///
+    pub NoUnknownTypeSelector {
+        version: "next",
+        name: "noUnknownTypeSelector",
+        language: "css",
+        recommended: false,
```
Comment (togami2864) [inline crates/biome_css_analyze/src/lint/nursery/no_unknown_type_selector.rs:54]:
```suggestion
        recommended: true,
        sources: &[RuleSource::Stylelint("selector-type-no-unknown")],
```

File: crates/biome_css_analyze/src/lint/nursery/no_unknown_type_selector.rs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+                rule_category!(),
+                span,
+                markup! {
+                    "Unknown type selector is not allowed."
+                },
+            )
+            .note(markup! {
+                "See "<Hyperlink href="https://developer.mozilla.org/en-US/docs/Web/CSS/Type_selectors">"MDN web docs"</Hyperlink>" for more details."
+            }).note(markup! {
+                "To resolve this issue, replace the unknown type selector with valid one."
```
Comment (togami2864) [inline crates/biome_css_analyze/src/lint/nursery/no_unknown_type_selector.rs:91]:
```suggestion
                "Consider replacing the unknown type selector with valid one."
```

File: crates/biome_css_analyze/src/utils.rs [hunk 3]
```diff
@@ -174,3 +174,15 @@ pub fn is_media_feature_name(prop: &str) -> bool {
     }
     false
 }
+
+/// Check if the input string is a known type selector.
+pub fn is_known_type_selector(prop: &str) -> bool {
+    let input = prop.to_lowercase();
+    fn is_custom_element(prop: &str) -> bool {
+        prop.contains('-') && prop.eq(prop.to_lowercase().as_str())
+    }
```
Comment (togami2864) [inline crates/biome_css_analyze/src/utils.rs:183]: You should define this outside of the function.

**Review by togami2864 — COMMENTED** [review]
Thank you!

File: crates/biome_css_analyze/src/keywords.rs
Comment (Mouvedia) [inline crates/biome_css_analyze/src/keywords.rs:1]: There are a lot of missing ones.
e.g. `nobr`, `model`, `strike`, `glyphRef`, etc.

check https://github.com/stylelint/stylelint/blob/main/lib/rules/selector-type-no-unknown/index.mjs#L94-L97 and https://github.com/stylelint/stylelint/blob/main/lib/reference/selectors.mjs#L47

**Comment by ematipico** [comment]
Hey @Kazuhiro-Mimaki, are you still interested in this PR?

**Comment by Kazuhiro-Mimaki** [comment]
I'm sorry but I forgot this issue... 🙇‍♂️ 
I'm still interested in and I'll do tomorrow.

File: crates/biome_css_analyze/src/keywords.rs [hunk 4]
```diff
@@ -5433,17 +5433,290 @@ pub const RESET_TO_INITIAL_PROPERTIES_BY_FONT: [&str; 13] = [
     "font-variation-settings",
 ];
 
+// https://developer.mozilla.org/ja/docs/Web/HTML/Element
+// https://github.com/sindresorhus/html-tags/blob/main/html-tags.json
+pub const HTML_TAGS: [&str; 146] = [
```
Comment (Mouvedia) [inline crates/biome_css_analyze/src/keywords.rs:5438]: These are missing: `big`, `bgsound`

Comment (Kazuhiro-Mimaki) [reply crates/biome_css_analyze/src/keywords.rs]: added

File: crates/biome_css_analyze/src/keywords.rs [hunk 5]
```diff
@@ -5433,17 +5433,292 @@ pub const RESET_TO_INITIAL_PROPERTIES_BY_FONT: [&str; 13] = [
...
+    "var",
+    "video",
+    "wbr",
+    "xmp",
+];
+
+// https://developer.mozilla.org/ja/docs/Web/SVG/Element
+// https://github.com/element-io/svg-tags/blob/master/lib/svg-tags.json
+pub const SVG_TAGS: [&str; 81] = [
+    "a",
```
Comment (Mouvedia) [inline crates/biome_css_analyze/src/keywords.rs:5592]: question

The `pub` seems to imply that is intended to be used by other rules.
If not, would it be better to remove the duplicates?
i.e. intersection with `HTML_TAGS`

File: crates/biome_css_analyze/src/keywords.rs [hunk 6]
```diff
@@ -5433,17 +5433,292 @@ pub const RESET_TO_INITIAL_PROPERTIES_BY_FONT: [&str; 13] = [
...
+    "font-face",
+    "font-face-format",
+    "font-face-name",
+    "font-face-src",
+    "font-face-uri",
+    "foreignObject",
+    "g",
+    "glyph",
+    "glyphRef",
+    "hatchpath",
```
Comment (Mouvedia) [inline crates/biome_css_analyze/src/keywords.rs:5643]: `hatch` is missing

File: crates/biome_css_analyze/src/keywords.rs (hunk 5 above)
Comment (ematipico) [reply crates/biome_css_analyze/src/keywords.rs:5592]: Good catch. It should be `pub(crate)` as we don't want to expose it for now

Comment (Kazuhiro-Mimaki) [reply crates/biome_css_analyze/src/keywords.rs:5592]: fixed
https://github.com/biomejs/biome/pull/2871/commits/13b6a1dd40f836630cc820d0ee956e680f2455f4

File: crates/biome_css_analyze/src/keywords.rs [hunk 7]
```diff
@@ -5433,17 +5433,293 @@ pub const RESET_TO_INITIAL_PROPERTIES_BY_FONT: [&str; 13] = [
...
+    "col",
+    "colgroup",
+    "content",
+    "data",
+    "datalist",
+    "dd",
+    "del",
+    "details",
+    "dfn",
+    "dir",
```
Comment (Mouvedia) [inline crates/biome_css_analyze/src/keywords.rs:5474]: `dialog` is missing

ref https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dialog

**Review by Mouvedia — CHANGES_REQUESTED** [review]
~~`dialog` is missing

ref https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dialog~~

File: crates/biome_css_analyze/src/keywords.rs [hunk 8]
```diff
@@ -5433,17 +5433,293 @@ pub const RESET_TO_INITIAL_PROPERTIES_BY_FONT: [&str; 13] = [
...
+    "progress",
+    "q",
+    "rb",
+    "rp",
+    "rt",
+    "rtc",
+    "ruby",
+    "s",
+    "samp",
+    "script",
```
Comment (Mouvedia) [inline crates/biome_css_analyze/src/keywords.rs:5551]: `shadow` is missing.

<hr />

Here's some browser history if you are curious:

- Firefox 29 added support for `<shadow>` behind the `dom.webcomponents.enabled` flag in late 2013.
- The tag was removed from the specification in 2015.

ref
https://bugzilla.mozilla.org/show_bug.cgi?id=887538
https://developer.mozilla.org/en-US/docs/Mozilla/Firefox/Releases/29
https://web.archive.org/web/20150429040731/http://w3c.github.io/webcomponents/spec/shadow/#the-shadow-element

File: crates/biome_css_analyze/src/keywords.rs
Comment (Mouvedia) [inline crates/biome_css_analyze/src/keywords.rs:1]: The current lists seem complete to me, now.
If you wanna support [VML](https://www.w3.org/TR/NOTE-VML), you could add another list.
i.e. for `shapetype`, `shape`, `group`

I don't think that's necessary though.

**Comment by ematipico** [comment]
Hi @Kazuhiro-Mimaki, I rebased your PR. Would you mind addressing the latest comments so we can merge them?

File: crates/biome_css_analyze/src/keywords.rs
Comment (Kazuhiro-Mimaki) [reply crates/biome_css_analyze/src/keywords.rs:1]: Thank you but I don't address this in this PR 🙏

**Review by ematipico — APPROVED** [review]

**Pull request merged**